tun/tun_illumos.go

```diff
@@ -0,0 +1,355 @@
+//go:build illumos && amd64
+// +build illumos,amd64
+
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright (C) 2017-2021 WireGuard LLC. All Rights Reserved.
+ * Copyright 2019 Joyent, Inc.
+ */
+
+package tun
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+const (
+	tunNode = "/dev/tun"
+)
+
+type NativeTun struct {
+	fd      int
+	name    string
+	tunFile *os.File
+	events  chan Event
+	errors  chan error
+}
+
+// XXX This is largely copied from the freebsd code...
+// XXX How do I test if it is actually working?
```

Comment:
Can you test this all and remove the XXX comments before I do a full review? If you have questions or need help with certain parts, feel free to say so in ordinary comments.